torch/csrc/autograd/profiler.cpp

```diff
@@ -16,35 +16,55 @@ TORCH_API void registerCUDAMethods(CUDAStubs* stubs) {
   cuda_stubs = stubs;
 }
 
-ProfilerState state = ProfilerState::Disabled;
-uint16_t next_thread_id = 0;
-std::mutex all_event_lists_mutex;
-std::list<std::shared_ptr<RangeEventList>> all_event_lists;
+// Assign all threads contiguous integer IDs so we can keep track of them
+std::atomic<int> next_thread_id{0};
+thread_local int thread_id = -1;
+
+thread_local std::shared_ptr<ProfilerInvocationState> invocation_state =
+    std::make_shared<ProfilerInvocationState>();
+
 thread_local std::shared_ptr<RangeEventList> event_list;
-thread_local uint16_t thread_id;
+thread_local ProfilerInvocationState* associated_invocation_state = invocation_state.get();
+
+WorkerPushProfileState::WorkerPushProfileState(std::shared_ptr<ProfilerInvocationState> st) {
+  old_state = invocation_state;
+  invocation_state = st;
+}
+
+WorkerPushProfileState::~WorkerPushProfileState() {
+  invocation_state = old_state;
+}
+
+std::shared_ptr<ProfilerInvocationState> currState() {
+  return invocation_state;
+}
 
 RangeEventList& getEventList() {
-  if (!event_list) {
-    std::lock_guard<std::mutex> guard(all_event_lists_mutex);
-    event_list = std::make_shared<RangeEventList>();
-    thread_id = next_thread_id++;
-    all_event_lists.emplace_front(event_list);
+  if (!event_list || associated_invocation_state != invocation_state.get()) {
+    std::lock_guard<std::mutex> guard(invocation_state->mutex);
+    if (thread_id == -1) {
+      thread_id = next_thread_id++;
+    }
+    if (!invocation_state->all_event_lists[thread_id]) {
+      event_list = invocation_state->all_event_lists[thread_id] = std::make_shared<RangeEventList>();
```

Comment:
if event_list is a local cache - don't you need to update it here even if no new list needs to be created?

Reply:
`event_list` is a pointer into the list owned by the `ProfilerInvocationState`. Only the thread-local state needs to be updated to point to the right thing